   

## AIM: Reconstruct ancestral genomes

Testing output of different programs to see if we can reconstruct fungal genome ancestor sensibly

### DATA

MLGO output of gene order
Progressive Cactus from alignment data


NOTE: There seem to be different trees for cactus and mglo?
cactus:(((Eam:1,Ebr:1)Anc3:1,Efe:1)Anc1:1,((Eel728:1,Eel732:1)Anc4:1,(Etp:1,(Ecl:1,Ety:1)Anc6:1)Anc5:1)Anc2:1)Anc0;
mglo:((Eam708,EfeFl1)A1,((Ebr1,(Eel728,Eel732)A2)A3,(Etp76,(Ety1756,Ecl1605)A4)A5)A6)A7;

#### Cactus data
DATA from Amelie
Export the HAL alignment as MAF referenced at the root
```
hal2maf cactus.hal cactus.maf
```
#### MGLO data
DATA from Murray Cox:
+ ortholog files
  1 The gene order in modern species is in ‘gene_order.txt’ 
  2	the phylogeny was fixed from standard phylogeny methods ‘gene_order.phy’
+ MGLO files
- geneorder.out
- gene_order.dat
geneorder.out file has lists of gene orders for ancestral genomes. 
ancestral genome A6 is the genome of interest

### Checked the distribution of E.clarkii chromosome 2 genes on Ancestral genome 6 (root of tree)
(in test_MLGO_gene_distributions)
Expect that it should be contiguous on the ancestor based on distribution at the tips
First check how many of the genes on each E.clarkii chromosomes end up on each ancestral chromosome
```
perl check_mglo_output_overlaps.pl data/Ecl_gene_order.dat data/A6_geneorder.out > report/Ecl_chr_distribution_on_A6_genome.txt
```
Graph data to see how they compare across tip genomes
```
R slave --vanilla < graph_distribution_query_on_target

```

### Checked the contiguous blocks of E.clarkii chromosome 2 genes on all tip and ancestral genomes
(in test_MLGO_gene_distributions)
Looked at the presence/absence (1/0) of E.clarkii chr 2 genes on each tip and ancestral genome.
(Note that this did not include checking the gene order, just pres/abs).

```
# tips
perl flag_gene_blocks.pl data/Ecl_chr2.txt data/Ecl_gene_order.dat report/Epichloe_flagged_with_Ecl_chr2.txt
# ancestors
perl flag_gene_blocks.pl data/Ecl_chr2.txt data/geneorder.out report/Ancestors_flagged_with_Ecl_chr2.txt
```

Results look pretty framented. I would expect A2 and A4 to have complete Ecl chr2s, but they don't seem to? However, A6 has genes on chrs 1,4,5 as per David's graphs, but they are not in clean blocks on those chromosomes.

### Check that genes on cactus alignment blocks are in the same order as MLGO ancestors

......





### Try CHRONicle 
GBK file conversion doesn't work with CHRONicle scripts (see archive_setup_gbk_CHROnicle.sh  )
Wrote scripts to convert files manually (def_from_gff.pl  make_prt.pl)
ran a bash script to implement them setup_chronicle.sh
Note that annotation currently only highlights genes and CHROnicle can deal with other genome features too
Note that the annotation files are not all there and don't have a standardised format - talk to David

#### Issues: 
##### 1SynChro falls over after the first (blast hurdle)
 - tried having mito chr labeled as '8'
 - tried having equal numbers of genes in prt and def files (see David about discrepancies)
 - tried putting def file in order of chr and position
##### 1Synchro returning a key error in first synteny block calculation
 - checked that all gene names in each of the def files are unique
 -  


